src/lib/reasoners/fun_sat.ml

```diff
@@ -177,6 +177,8 @@ module Make (Th : Theory.S) = struct
     unit_facts_cache : (E.gformula * Ex.t) ME.t ref;
     last_saved_model : Models.t Lazy.t option ref;
     unknown_reason : Sat_solver_sig.unknown_reason option;
+    declare_ids : Id.typed Vec.t;
+    declare_lim : int Vec.t;
```

Comment:
This was my first implementation and it didn't work. A functional solution would be to use a list of lists:
```ocaml 
declare_stack: Id.typed list list
```
We have to `pop/push`  this stack in `Fun_sat.pop/Fun_sat.push`. 
I didn't adopt this solution because I afraid that a list of lists isn't cache-friendly. I understand that you want to keep the environment of `FunSAT` as pure as possible.

Honestly, I believe we should pull up all the imperative stuff of `FunSAT` into `Funsat_frontend` and use a stack of environment of `FunSAT` to simulate the push and the pop functions. In this design, we'll use a vector for declarations in the imperative environment of `Funsat_frontend`.

Reply:
Ah, right, `push` and `pop` forces us to be imperative. Oh well. I guess the way to do this while staying consistent with the current `Fun_sat` module is `Id.typed list Stack.t` rather than `Id.typed list list` then.

> I didn't adopt this solution because I afraid that a list of lists isn't cache-friendly.

You did write too much C++ :) A list of lists (or a stack of lists, which is the same thing) is very probably not going to be an issue here because we never actually look at the lists until model generation (and also because this only impacts calls to `push` and `pop`, which is virtually 0% of Alt-Ergo's runtime).

> In this design, we'll use a vector for declarations in the imperative environment of Funsat_frontend.

If we use `Funsat_frontend` for `push` and `pop` we don't need a separate vector for declarations in `Funsat_frontend`, we can simply have `declare_ids : Id.typed list` in `Funsat` which will get tracked by the main stack in the frontend, I believe.

Reply:
I agree to use `Id.typed list Stack.t` :)